packages/gatsby/cache-dir/json-store.js

```diff
@@ -6,81 +6,87 @@ import socketIo from "./socketIo"
 import omit from "lodash/omit"
 import get from "lodash/get"
 
+const getPathFromProps = props => {
+  if (props.isPage) {
+    return get(props.pageResources, `page.path`)
+  } else {
+    return `/dev-404-page/`
+  }
+}
+
 class JSONStore extends React.Component {
   constructor(props) {
     super(props)
     this.state = {
-      data: {},
+      data: null,
+      path: null,
     }
 
-    this.socket = socketIo()
-
     this.setPageData = this.setPageData.bind(this)
-    this.getPageData = this.getPageData.bind(this)
-  }
 
-  componentDidMount() {
+    this.socket = socketIo()
     this.socket.on(`queryResult`, this.setPageData)
   }
 
-  shouldComponentUpdate(nextProps, nextState) {
-    if (nextProps !== this.props) return true
-
-    // if json for nextState is not available
-    const nextJsonId = get(nextProps.pageResources, `page.jsonName`)
-    if (!nextState.data[nextJsonId]) return false
-
-    // if nextState json is the same as current state json
-    const sameDataPath =
-      get(nextState, `data[${nextJsonId}].dataPath`) ===
-      get(this, `state.data[${nextJsonId}].dataPath`)
+  componentWillMount() {
+    this.registerPath(getPathFromProps(this.props))
+  }
 
-    if (sameDataPath) return false
+  componentWillReceiveProps(nextProps) {
+    const { path } = this.state
+    const newPath = getPathFromProps(nextProps)
 
-    return true
+    if (path !== newPath) {
+      this.unregisterPath(path)
+      this.registerPath(newPath)
+    }
   }
 
-  setPageData(newData) {
-    this.setState({
-      data: { [newData.path]: newData },
-    })
+  registerPath(path) {
+    this.setState({ path })
+    this.socket.emit(`registerPath`, path)
   }
 
-  getPageData(path) {
-    const res = this.state.data[path]
+  unregisterPath(path) {
+    this.setState({ data: null, path: null })
+    this.socket.emit(`unregisterPath`, path)
+  }
 
-    // always check for fresh data
-    this.socket.emit(`getPageData`, path)
+  componentWillUnmount() {
+    this.unregisterPath(this.state.path)
+  }
 
-    if (!res || !res.data) return false
-    return JSON.parse(res.data)
+  setPageData({ path, result }) {
+    if (this.state.path === path)
+      this.setState({
+        data: result,
+      })
   }
 
   render() {
     const { isPage, pages, pageResources } = this.props
+    const { data } = this.state
     const propsWithoutPages = omit(this.props, `pages`)
 
-    if (isPage) {
-      const jsonId = get(pageResources, `page.jsonName`)
-      const pageData = this.getPageData(jsonId)
-      if (pageData === false) return ``
+    if (!data) {
+      return null
```

Comment:
here we could display loader

Reply:
We need suspense for this!

Reply:
do You mean new async render stuff that Dan Abramov demoed last month?

Reply:
yeah :-)

we'll be able to rewrite all of our data/code loading to use that once it's launched